DFC.Composite.Shell/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -27,64 +39,263 @@ public static class ServiceCollectionExtensions
+                nameof(SitemapClientOptions),
+                policyOptions,
+                configuration);
+
+            AddHttpClientWithPolicies<IApplicationRobotService, ApplicationRobotService, RobotClientOptions>(
+                services,
+                policyRegistry,
+                $"{nameof(RobotClientOptions)}_{nameof(PolicyOptions.HttpRetry)}",
+                $"{nameof(RobotClientOptions)}_{nameof(PolicyOptions.HttpCircuitBreaker)}",
+                nameof(RobotClientOptions),
+                policyOptions,
+                configuration);
+
+            AddHttpClientWithPolicies<IBannerService, BannerService, BannerClientOptions>(
+               services,
+               policyRegistry,
+               $"{nameof(BannerClientOptions)}_{nameof(PolicyOptions.HttpRetry)}",
+               $"{nameof(BannerClientOptions)}_{nameof(PolicyOptions.HttpCircuitBreaker)}",
+               nameof(BannerClientOptions),
+               policyOptions,
+               configuration);
+        }
+
+        private static bool AppRegistryRequestRegistered(this IServiceCollection services)
+        {
+            return services.Any(service => service.ServiceType == typeof(IAppRegistryService));
+        }
+
+        private static void AddApplicationClientHttp(
+            this IServiceCollection services,
+            IConfiguration configuration,
+            PolicyOptions policyOptions,
+            IPolicyRegistry<string> policyRegistry)
+        {
+            var registeredUrls = GetRegisteredUrls(services);
+
+            if (registeredUrls == null)
+            {
+                registeredUrls = new RegisteredUrls(default);
+            }
+
+            services.AddSingleton<IRegisteredUrls>(registeredUrls);
+
+            foreach (var registeredUrl in registeredUrls.GetAll())
+            {
+                AddHttpClientWithPolicies<IUriSpecifcHttpClientFactory, UriSpecifcHttpClientFactory, ApplicationClientOptions>(
+                    services,
+                    policyRegistry,
+                    $"{registeredUrl}_{nameof(ApplicationClientOptions)}_{nameof(PolicyOptions.HttpRetry)}",
+                    $"{registeredUrl}_{nameof(ApplicationClientOptions)}_{nameof(PolicyOptions.HttpCircuitBreaker)}",
+                    nameof(ApplicationClientOptions),
+                    policyOptions,
+                    configuration,
+                    $"{registeredUrl}_{nameof(UriSpecifcHttpClientFactory)}")
+                .AddHttpMessageHandler<CompositeSessionIdDelegatingHandler>()
+                .AddHttpMessageHandler<CookieDelegatingHandler>();
+            }
+        }
+
+        private static RegisteredUrls GetRegisteredUrls(IServiceCollection services)
+        {
+            var urls = Task.Run(() =>
+            {
+                var serviceProvider = services.BuildServiceProvider();
+                return serviceProvider.GetService<IAppRegistryDataService>().GetAppRegistrationModels();
+            })
```

Comment:
No need for Task.Run here either. creating a new thread and blocking this one by using .Result is highly inefficient.

Reply:
I'm thinking back to Framework days I guess - either way its only done once, but i'll remove that bit

Reply:
Changed now